File: src/test_polars/utils/process_data.py
"""Module contains functions for joining dataframes."""

import polars as pl


def join_inner(df1: pl.DataFrame, df2: pl.DataFrame) -> pl.DataFrame:
    """
    Perform an inner join on two dataframes.

    Parameters
    ----------
    df1: polars.DataFrame
        The left DataFrame to join.
    df2: polars.DataFrame
        The right DataFrame to join.

    Returns
    -------
    polars.DataFrame
        The joined DataFrame.

    """
    return df1.join(
        df2,
        left_on=["src_ip", "dst_ip", "seq_num"],
        right_on=["dst_ip", "src_ip", "seq_num"],
        how="inner",
    )


def join_right(df1: pl.DataFrame, df2: pl.DataFrame) -> pl.DataFrame:
    """
    Perform a right join on two dataframes.

    Parameters
    ----------
    df1: polars.DataFrame
        The left DataFrame to join.
    df2: polars.DataFrame
        The right DataFrame to join.

    Returns
    -------
    polars.DataFrame
        The joined DataFrame.

    """
    return df1.join(
        df2,
        left_on=["src_ip", "dst_ip", "seq_num"],
        right_on=["dst_ip", "src_ip", "seq_num"],
        how="right",
    )


def join_left(df1: pl.DataFrame, df2: pl.DataFrame) -> pl.DataFrame:
    """
    Perform a left join on two dataframes.

    Parameters
    ----------
    df1: polars.DataFrame
        The left DataFrame to join.
    df2: polars.DataFrame
        The right DataFrame to join.

    Returns
    -------
    polars.DataFrame
        The joined DataFrame.

    """
    return df1.join(
        df2,
        left_on=["src_ip", "dst_ip", "seq_num"],
        right_on=["dst_ip", "src_ip", "seq_num"],
        how="left",
    )


def join_full(df1: pl.DataFrame, df2: pl.DataFrame) -> pl.DataFrame:
    """
    Perform a full join on two dataframes.

    Parameters
    ----------
    df1: polars.DataFrame
        The left DataFrame to join.
    df2: polars.DataFrame
        The right DataFrame to join.

    Returns
    -------
    polars.DataFrame
        The joined DataFrame.

    """
    return df1.join(
        df2,
        left_on=["src_ip", "dst_ip", "seq_num"],
        right_on=["dst_ip", "src_ip", "seq_num"],
        how="full",
    )


def join_outer(df1: pl.DataFrame, df2: pl.DataFrame) -> pl.DataFrame:
    """
    Perform an outer join on two dataframes.

    Parameters
    ----------
    df1: polars.DataFrame
        The left DataFrame to join.
    df2: polars.DataFrame
        The right DataFrame to join.

    Returns
    -------
    polars.DataFrame
        The joined DataFrame.

    """
    return df1.join(
        df2,
        left_on=["src_ip", "dst_ip", "seq_num"],
        right_on=["dst_ip", "src_ip", "seq_num"],
        how="outer",
    )


def join_semi(df1: pl.DataFrame, df2: pl.DataFrame) -> pl.DataFrame:
    """
    Perform a semi join on two dataframes.

    Parameters
    ----------
    df1: polars.DataFrame
        The left DataFrame to join.
    df2: polars.DataFrame
        The right DataFrame to join.

    Returns
    -------
    polars.DataFrame
        The joined DataFrame.

    """
    return df1.join(
        df2,
        left_on=["src_ip", "dst_ip", "seq_num"],
        right_on=["dst_ip", "src_ip", "seq_num"],
        how="semi",
    )


def join_anti(df1: pl.DataFrame, df2: pl.DataFrame) -> pl.DataFrame:
    """
    Perform an anti join on two dataframes.

    Parameters
    ----------
    df1: polars.DataFrame
        The left DataFrame to join.
    df2: polars.DataFrame
        The right DataFrame to join.

    Returns
    -------
    polars.DataFrame
        The joined DataFrame.

    """
    return df1.join(
        df2,
        left_on=["src_ip", "dst_ip", "seq_num"],
        right_on=["dst_ip", "src_ip", "seq_num"],
        how="anti",
    )

File: src/utils/process_data.rs
use polars::prelude::{DataFrame, DataFrameJoinOps, JoinArgs, JoinType};

/// Perform an inner join on two dataframes.
///
/// # Errors
///
/// If the join fails (because the column names do not exist, for example), an empty
/// dataframe is returned.
pub fn join_inner(df1: &DataFrame, df2: &DataFrame) -> DataFrame {
    df1.join(
        df2,
        ["src_ip", "dst_ip", "seq_num"],
        ["dst_ip", "src_ip", "seq_num"],
        JoinArgs::new(JoinType::Inner),
        None,
    )
    .unwrap_or(DataFrame::default())
}

/// Perform a right join on two dataframes.
///
/// # Errors
///
/// If the join fails (because the column names do not exist, for example), an empty
/// dataframe is returned.
pub fn join_right(df1: &DataFrame, df2: &DataFrame) -> DataFrame {
    df1.join(
        df2,
        ["src_ip", "dst_ip", "seq_num"],
        ["dst_ip", "src_ip", "seq_num"],
        JoinArgs::new(JoinType::Right),
        None,
    )
    .unwrap_or(DataFrame::default())
}

/// Perform a left join on two dataframes.
///
/// # Errors
///
/// If the join fails (because the column names do not exist, for example), an empty
/// dataframe is returned.
pub fn join_left(df1: &DataFrame, df2: &DataFrame) -> DataFrame {
    df1.join(
        df2,
        ["src_ip", "dst_ip", "seq_num"],
        ["dst_ip", "src_ip", "seq_num"],
        JoinArgs::new(JoinType::Left),
        None,
    )
    .unwrap_or(DataFrame::default())
}

/// Perform a full join on two dataframes.
///
/// # Errors
///
/// If the join fails (because the column names do not exist, for example), an empty
/// dataframe is returned.
pub fn join_full(df1: &DataFrame, df2: &DataFrame) -> DataFrame {
    df1.join(
        df2,
        ["src_ip", "dst_ip", "seq_num"],
        ["dst_ip", "src_ip", "seq_num"],
        JoinArgs::new(JoinType::Full),
        None,
    )
    .unwrap_or(DataFrame::default())
}

/// Perform an outer join on two dataframes.
///
/// # Errors
///
/// If the join fails (because the column names do not exist, for example), an empty
/// dataframe is returned.
pub fn join_outer(df1: &DataFrame, df2: &DataFrame) -> DataFrame {
    df1.join(
        df2,
        ["src_ip", "dst_ip", "seq_num"],
        ["dst_ip", "src_ip", "seq_num"],
        JoinArgs::new(JoinType::Full),
        None,
    )
    .unwrap_or(DataFrame::default())
}

/// Perform a semi join on two dataframes.
///
/// # Errors
///
/// If the join fails (because the column names do not exist, for example), an empty
/// dataframe is returned.
pub fn join_semi(df1: &DataFrame, df2: &DataFrame) -> DataFrame {
    df1.join(
        df2,
        ["src_ip", "dst_ip", "seq_num"],
        ["dst_ip", "src_ip", "seq_num"],
        JoinArgs::new(JoinType::Semi),
        None,
    )
    .unwrap_or(DataFrame::default())
}

/// Perform an anti join on two dataframes.
///
/// # Errors
///
/// If the join fails (because the column names do not exist, for example), an empty
/// dataframe is returned.
pub fn join_anti(df1: &DataFrame, df2: &DataFrame) -> DataFrame {
    df1.join(
        df2,
        ["src_ip", "dst_ip", "seq_num"],
        ["dst_ip", "src_ip", "seq_num"],
        JoinArgs::new(JoinType::Anti),
        None,
    )
    .unwrap_or(DataFrame::default())
}
